### PR TITLE
Add feature amount calculation to feature shapefile upload [MRXN-468]

### PR DESCRIPTION
**Pull request by KevSanchez**
### Overview

_Please write a description. If the PR is hard to understand, provide a quick explanation of the code._

### Designs

_Link to the related design prototypes (if applicable)_

### Testing instructions

_Please explain how to test the PR: ID of a dataset, steps to reach the feature, etc._

### Feature relevant tickets

_Link to the related task manager tickets_

---

## Checklist before submitting

- [ ] Meaningful commits and code rebased on `develop`.
- [ ] If this PR adds feature that should be tested for regressions when
      deploying to staging/production, please add brief testing instructions
      to the deploy checklist (`docs/deployment-checklist.md`)
- [ ] Update CHANGELOG file

### CODE REVIEW

File: api/apps/api/src/modules/geo-features/geo-features.service.ts [hunk 1]
```diff
@@ -431,6 +452,24 @@ export class GeoFeaturesService extends AppBaseService<
...
+    );
+    await repo.save(
+      featureAmounts.map(({ amount, projectPuId, featureId }) => ({
+        projectId,
+        featureId,
+        amount,
+        projectPuId,
+      })),
+    );
+  }
```
Comment (hotzevzl) [inline api/apps/api/src/modules/geo-features/geo-features.service.ts:471]: I think this may break under pressure, aka when inserting a high enough number of PU amounts at once so that the combined number of query parameters goes above PostgreSQL's limit for this.


```suggestion
// add to imports
import { CHUNK_SIZE_FOR_BATCH_GEODB_OPERATIONS } from '@marxan-geoprocessing/utils/chunk-size-for-batch-geodb-operations';
//and here
    await repo.save(
      featureAmounts.map(({ amount, projectPuId, featureId }) => ({
        projectId,
        featureId,
        amount,
        projectPuId,
      })),
      {
        chunk: CHUNK_SIZE_FOR_BATCH_GEODB_OPERATIONS,
      }
    );
  }
```

File: api/libs/feature-amounts-per-planning-unit/src/feature-amounts-per-planning-units.service.ts [hunk 2]
```diff
@@ -41,48 +41,51 @@ export class FeatureAmountsPerPlanningUnitService {
...
     /**
      * @TODO further performance savings: limiting scans to planning_units_geom
      * by partition (we need to get the grid shape from the parent project); use
      * && operator instead of st_intersects() for bbox-based calculation of
      * intersections.
      */
+    geoEntityManager = geoEntityManager
+      ? geoEntityManager
+      : this.geoEntityManager;
+
```
Comment (hotzevzl) [inline api/libs/feature-amounts-per-planning-unit/src/feature-amounts-per-planning-units.service.ts:56]: I think you could simplify this:

```suggestion
    geoEntityManager: EntityManager = this.geoEntityManager,
  ): Promise<ComputeFeatureAmountPerPlanningUnit[]> {
    /**
     * @TODO further performance savings: limiting scans to planning_units_geom
     * by partition (we need to get the grid shape from the parent project); use
     * && operator instead of st_intersects() for bbox-based calculation of
     * intersections.
     */

```

**Review by hotzevzl — COMMENTED** [review]
thanks @KevSanchez! I'm still trying out the functionality in my local dev environment, but in the meanwhile, please see a couple of notes inline.

File: api/apps/api/src/modules/geo-features/geo-features.service.ts (hunk 1 above)
Comment (hotzevzl) [reply api/apps/api/src/modules/geo-features/geo-features.service.ts]: 🤔 I thought I had posted a follow-up comment via my IDE but I think I may have messed that up - basically it was about my ill-informed advice to use `CHUNK_SIZE_FOR_BATCH_GEODB_OPERATIONS` as this is defined in the geoprocessing service and so it will not be importable directly in the api service. I had some random rants about semantics in my vanished comment, but basically tl;dr please let's use `CHUNK_SIZE_FOR_BATCH_APIDB_OPERATIONS` here without moving anything else around (moving constants to libs, etc.)

**Comment by hotzevzl** [comment]
@KevSanchez works beautifully for me - please have a look at the leftover inline comments including my own backpedaling on the suggestion to use `CHUNK_SIZE_FOR_BATCH_GEODB_OPERATIONS` and then we can merge ❤️

File: api/apps/api/src/modules/geo-features/geo-features.service.ts [hunk 3]
```diff
@@ -854,6 +893,10 @@ export class GeoFeaturesService extends AppBaseService<
       .groupBy('fappu.feature_id')
       .getRawMany();
 
+    if (minAndMaxAmountsForFeatures.length === 0) {
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (KevSanchez) [reply api/apps/api/src/modules/geo-features/geo-features.service.ts:900]: Yeah I agree, it would be better to throw error, and catch it later. 
The only issue in this case is that by the time this function executes, where it gets executed, the changes relater to feature amounts have already been committed (geo-features.service.ts line 437 and features-amount-upload.service.ts line 116), and catching the error won't do much more than logging it (which already better than nothing, indeed)

**Review by hotzevzl — APPROVED** [review]
let's merge once all is green, thanks! 🚀
